### PR TITLE
Added atreides-01a

### PR DESCRIPTION
**Pull request by abcdefg30**
Three things left to do:
- [x] Add a mission preview
- [x] Have different difficulties (ATM selecting one won't change anything ingame)
- [x] Update the description

### CODE REVIEW

**Comment by penev92** [comment]
Here are some more:
 - [ ] I can't seem to fail the mission.
 - [ ] You should copy the "Objective: ..." part from the map description and use that as the primary objective.
 - [ ] In my opinion using #7156 for the text messages would look better (and closer to the original), but this is optional. You can even drop "Mentat:".

**Comment by abcdefg30** [comment]
> In my opinion using #7156 for the text messages would look better (and closer to the original), but this is optional. You can even drop "Mentat:".

As "discussed" on IRC #7156 is not what you actually wanted.

Everything else is done, except:
> Have different difficulties (ATM selecting one won't change anything ingame)

**Comment by Unit158** [comment]
Difficulties were implemented in the original by making things cheaper or more expensive. I suggest we follow through with this instead of adding our own twists atm.

**Comment by Mailaender** [comment]
Difficulties (easy, easier, even easier?) and adding the .lua files to `OpenRA.sln` (especially if this is impossible with your IDE) aren't important right now imho.

**Comment by penev92** [comment]
True, prices change with difficulty in the original (I don't know about anything else), but making stuff more expensive in no way enhances the gameplay of the mission - it just makes it finish slower, which in turn actually makes it just tedious.
This approach will have its merit in later missions, considering how Dune 2000 is very economy-oriented, but for this mission it is pointless. Either remove difficulties, or go back to the original idea of using them to make it more interesting.

**Comment by pchote** [comment]
I vote for removing difficulties until we have something interesting to do with them.

**Comment by abcdefg30** [comment]
Updated with difficulties.

**Comment by abcdefg30** [comment]
Fixed `OpenRA.sln`.

File: OpenRA.sln [hunk 1]
```diff
@@ -59,6 +59,11 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Red Alert Lua scripts", "Re
 		mods\ra\maps\survival02\survival02.lua = mods\ra\maps\survival02\survival02.lua
 	EndProjectSection
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Dune 2000 Lua scripts", "Dune 2000 Lua scripts", "{06B1AE07-DDB0-4287-8700-A8CD9A0E652E}"
+	ProjectSection(SolutionItems) = preProject
+		mods\d2k\maps\atreides-01a\atreides01a.lua = mods\d2k\maps\atreides-01a\atreides01a.lua
```
Comment (Mailaender) [inline OpenRA.sln]: You can add https://github.com/OpenRA/OpenRA/blob/bleed/mods/d2k/maps/shellmap/shellmap.lua here, too while you are at it. =)

**Comment by penev92** [comment]
OK, let's see.
 - Easy is pretty much the same as the original, except the bounties from killing enemies. But there are only 6 of them, and you get 10 "solaris" a piece, so that doesn't really matter.
 - Medium I have not tried.
 - Hard is setting a second worm spawner, which can do anything from make your life easier (the worms do attack Harkonnen...) to making it very hard (your "safe" harvesting place isn't safe any more). The attacks are much more than I expected, but IMO make the mission more interesting. Completing the secondary objective is almost impossible if you don't know how the script works though.

**Comment by abcdefg30** [comment]
Updated.

**Comment by Phrohdoh** [comment]
In the secondary objective `units and reinforcements` could become `forces` for brevity.

File: mods/d2k/maps/atreides-01a/atreides01a.lua [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+		end
+	end)
+end
+
+InitObjectives = function()
+	Trigger.OnObjectiveAdded(player, function(p, id)
+		Media.DisplayMessage(p.GetObjectiveDescription(id), "New " .. string.lower(p.GetObjectiveType(id)) .. " objective")
+	end)
+
+	KillAtreides = harkonnen.AddPrimaryObjective("Kill all Atreides units.")
```
Comment (Phrohdoh) [inline mods/d2k/maps/atreides-01a/atreides01a.lua]: Is this necessary (for losing)?
Harkonnen aren't playable here, obviously.

File: mods/d2k/maps/atreides-01a/atreides01a.lua [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	GatherSpice = player.AddPrimaryObjective("Harvest " .. tostring(ToHarvest) .. " Solaris worth of Spice.")
+	KillHarkonnen = player.AddSecondaryObjective("Eliminate all Harkonnen units and reinforcements\n in the area.")
+
+	Trigger.OnObjectiveCompleted(player, function(p, id)
+		Media.DisplayMessage(p.GetObjectiveDescription(id), "Objective completed")
+	end)
+	Trigger.OnObjectiveFailed(player, function(p, id)
+		Media.DisplayMessage(p.GetObjectiveDescription(id), "Objective failed")
```
Comment (Phrohdoh) [inline mods/d2k/maps/atreides-01a/atreides01a.lua]: Could we get full-stops at the end of these messages?

Comment (abcdefg30) [inline mods/d2k/maps/atreides-01a/atreides01a.lua]: "Objective completed/failed" are only those prefixes.
So full-stops are at the end of the messages, as they are at the end of the objective descriptions.

File: mods/d2k/maps/atreides-01a/atreides01a.lua (hunk 2 above)
Comment (abcdefg30) [inline mods/d2k/maps/atreides-01a/atreides01a.lua]: Yes, it is used to make the mission fail properly.

File: OpenRA.sln (hunk 1 above)
Comment (abcdefg30) [inline OpenRA.sln]: No problem. :) Done with https://github.com/OpenRA/OpenRA/pull/8117#issuecomment-101396667.

File: mods/d2k/maps/atreides-01a/atreides01a.lua (hunk 3 above)
Comment (Phrohdoh) [inline mods/d2k/maps/atreides-01a/atreides01a.lua]: Okay. :smile:

**Comment by abcdefg30** [comment]
> In the secondary objective  units and reinforcements  could become  forces  for brevity.

I actually wanted to clarify that you have to wait for the reinforcements to complete this objective.

**Comment by penev92** [comment]
With some of the latest changes (now taking `player.Cash` into consideration) I can now sell my conyard the moment the mission starts and I get "Mission Accomplished".
I suspect the check in `Tick` is ran before
`Trigger.OnRemovedFromWorld(AtreidesConyard, function()`  has a chance to.

**Comment by abcdefg30** [comment]
Depends on #8162 now.

**Comment by Mailaender** [comment]
As https://github.com/OpenRA/OpenRA/pull/8162 is causing a controversy can you keep the rule change contained to this mission for now?

**Comment by abcdefg30** [comment]
> As #8162 is causing a controversy can you keep the rule change contained to this mission for now?

Actually this doesn't depend on #8162 anymore but on #8191 because https://github.com/OpenRA/OpenRA/pull/8162#issuecomment-102045966.

So this is the plan:
- Get #8191 merged.
- Add a `Harvested resources: x/2500` missionUI-text

**Comment by abcdefg30** [comment]
Updated with a `Harvested resources: x/2500` missionUI-text.

File: mods/d2k/maps/atreides-01a/map.yaml [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+		Location: 9,13
+		Owner: Atreides
+	Actor8: rifle
+		Location: 15,13
+		Owner: Atreides
+	Actor9: rifle
+		Location: 14,14
+		Owner: Atreides
+		Facing: 128
+		TurretFacing: 128
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-01a/map.yaml]: TurretFacing?

File: mods/d2k/maps/atreides-01a/atreides01a.lua [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+end
+
+InitObjectives = function()
+	Trigger.OnObjectiveAdded(player, function(p, id)
+		Media.DisplayMessage(p.GetObjectiveDescription(id), "New " .. string.lower(p.GetObjectiveType(id)) .. " objective")
+	end)
+
+	KillAtreides = harkonnen.AddPrimaryObjective("Kill all Atreides units.")
+	GatherSpice = player.AddPrimaryObjective("Harvest " .. tostring(ToHarvest) .. " Solaris worth of Spice.")
+	KillHarkonnen = player.AddSecondaryObjective("Eliminate all Harkonnen units and reinforcements\n in the area.")
```
Comment (obrakmann) [inline mods/d2k/maps/atreides-01a/atreides01a.lua]: Very minor nit: you can remove the space after the line break.

File: mods/d2k/maps/atreides-01a/map.yaml (hunk 4 above)
Comment (abcdefg30) [inline mods/d2k/maps/atreides-01a/map.yaml]: Weird. Removed it.

**Comment by obrakmann** [comment]
Looks good to me. :+1:

File: mods/d2k/maps/atreides-01a/atreides01a.lua (hunk 5 above)
Comment (abcdefg30) [inline mods/d2k/maps/atreides-01a/atreides01a.lua]: Removed it.

**Comment by abcdefg30** [comment]
Updated.

**Comment by penev92** [comment]
- Now that upgrades are in, you need to disable the ConYard upgrade for this mission.
 - Also on Hard the last Harkonnen come too late. I got "Mission complete" just as they were entering the map, which meant I couldn't finish the secondary objective.

**Comment by abcdefg30** [comment]
Disabled the upgrades queue.

**Comment by penev92** [comment]
Thanks! :+1:

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/3a1c52437fb5df4c79baca054a95a845fc81f378...063de23d9891c585661eef6dea22ce0d9c664d1d)

**Pull request merged**